certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,209 @@
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount > 0 => currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+    assert threshold <= guardianStorageContract.entries[safeContract].count;
+}
+
+// This integrity rule verifies that the guardian can always be revoked considering ideal conditions.
+rule guardianCanAlwaysBeRevoked(env e, address guardian, address prevGuardian, uint256 threshold) {
     requireSocialRecoveryModuleEnabled();
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // Guardian should not be zero address.
+    require guardian != 0;
+    // If new threshold is 0, then the guardian count should be 1.
+    require threshold == 0 => guardianStorageContract.entries[safeContract].count == 1;
+    // Else, if new threshold is greater than 0, then the guardian count should be greater than the threshold.
+    require threshold > 0 => guardianStorageContract.entries[safeContract].count > threshold;
+    // The address should be a guardian.
+    require currentContract.isGuardian(safeContract, guardian);
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+    require guardianStorageContract.entries[safeContract].guardians[prevGuardian] == guardian;
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;    
+
+    // Safe Contract should be the sender of the transaction.
+    require e.msg.sender == safeContract;
+    currentContract.revokeGuardianWithThreshold@withrevert(e, safeContract, prevGuardian, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian &&
+        !currentContract.isGuardian(safeContract, guardian) &&
+        currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies the possibilites in which the revocation of a new guardian can revert.
+rule revokeGuardianRevertPossibilities(env e, address otherAccount, address prevGuardian, uint256 threshold) {
+    bool isGuardian = currentContract.isGuardian(safeContract, otherAccount);
 
-    safeContract.disableModule@withrevert(e, prevModule, currentContract);
+    currentContract.revokeGuardianWithThreshold@withrevert(e, safeContract, prevGuardian, otherAccount, threshold);
     bool isReverted = lastReverted;
 
-    assert !isReverted => !safeContract.isModuleEnabled(currentContract);
+    assert isReverted =>
+        !isGuardian ||
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        otherAccount == 0 ||
+        otherAccount == SENTINEL() ||
+        guardianStorageContract.entries[safeContract].count == 0 ||
```

Comment:
If we require linked list integrity, then this isn't a possible condition for reverting (as it would imply `!isGuardian`).

Reply:
Attaching the issue where we should address this once the linked list integrity is ensured: #24

Reply:
My concrete ask is to remove the `guardianStorageContract.entries[safeContract].count == 0` condition (which may require you adding a `requireLinkedListIntegrity` to the rule).

I don't think this warrants a follow up issue.

Reply:
Done 👍🏾